pyfr/backends/cuda/cublaslt.py

```diff
@@ -0,0 +1,285 @@
+        if a.nrow != out.nrow or a.ncol != b.nrow or b.ncol != out.ncol:
+            raise ValueError('Incompatible matrices for out = a*b')
+
+        # CUBLASLt expects inputs to be column-major (or Fortran order in
+        # numpy parlance).  However as C = A*B => C^T = (A*B)^T
+        # = (B^T)*(A^T) with a little trickery we can multiply our
+        # row-major matrices directly.
+        m, n, k = b.ncol, a.nrow, a.ncol
+        A, B, C = b, a, out
+
+        # Cache key
+        ckey = (A.dtype, alpha, beta, m, n, k, A.leaddim, B.leaddim, C.leaddim)
+
+        # α and β factors for D = α*(A*B) + β*C
+        if a.dtype == np.float64:
+            alpha_ct, beta_ct = c_double(alpha), c_double(beta)
+            ctype = w.COMPUTE_64F
+            dtype = w.R_64F
+        else:
+            alpha_ct, beta_ct = c_float(alpha), c_float(beta)
+            ctype = w.COMPUTE_32F
+            dtype = w.R_32F
+
+        ws_ptr = c_void_p()
+        desc = GEMMDesc(None, None, None, None, None, 0)
+
+        # Convenience wrapper
+        def gemm(stream):
+            w.cublasLtMatmul(
+                h, desc.mm, byref(alpha_ct), A, desc.a, B,
+                desc.b, byref(beta_ct), C, desc.c, C, desc.c,
+                byref(desc.algo), ws_ptr, desc.ws_size, stream
+            )
+
+        # Obtain the performance of the kernel
+        try:
+            desc, dt = self._mul_cache[ckey]
+        except KeyError:
+            # Create matrix layouts
+            a_desc = CUBLASLtMatrixLayout(self, A, dtype)
+            b_desc = CUBLASLtMatrixLayout(self, B, dtype)
+            c_desc = CUBLASLtMatrixLayout(self, C, dtype)
+
+            # Matmul descriptor
+            mm_attr = [(w.MATMUL_DESC_TRANSA, w.OP_N, c_int),
+                       (w.MATMUL_DESC_TRANSB, w.OP_N, c_int),
+                       (w.MATMUL_DESC_TRANSC, w.OP_N, c_int)]
+            mm_desc = CUBLASLtMatmulDesc(self, ctype, dtype, mm_attr)
+
+            # Heuristic preference descriptor
+            pref_attr = [(w.MATMUL_PREF_MAX_WORKSPACE_BYTES,
+                          w.WORKSPACE_MAX_SIZE,
+                          c_size_t)]
+            pref = CUBLASLtMatmulPreference(self, pref_attr)
+
+            # Get matmul heuristics
+            heurs = (CUBLASLtMatmulHeuristicResult * self.nkerns)()
+            nreturn = c_int()
+            w.cublasLtMatmulAlgoGetHeuristic(
+                h, mm_desc, a_desc, b_desc, c_desc, c_desc, pref, self.nkerns,
+                heurs, byref(nreturn)
+            )
+
+            # Save a copy of the contents of the output matrix
+            out_np = getattr(out, 'parent', out).get()
+
+            # Allocate some temporary worspace for benchmarks
+            ws_ptr = CUDADevAlloc(cuda, w.WORKSPACE_MAX_SIZE)
+
+            # Benchmark the kernel
+            best_kern = None
+            for heur in heurs[:nreturn.value]:
+                # Allocate gemm workspace
+                ws_size = heur.workspaceSize
```

Comment:
Not sure this is needed?  Just pass in a larger size and I don't think it will care

Reply:
As we have the number, we might as well give it the right size.